astropy/io/fits/hdu/compressed/compressed.py

```diff
@@ -572,408 +526,146 @@ def data(self):
-            self.scale(BITPIX2DTYPE[self._orig_bitpix])
+            for _ in range(image_blanks - table_blanks):
+                self._tmp_bintable.header.append()
 
-        if self._has_data:
-            self._update_compressed_data()
-
-            # Use methods in the superclass to update the header with
-            # scale/checksum keywords based on the data type of the image data
-            self._update_pseudo_int_scale_keywords()
-
-            # Shove the image header and data into a new ImageHDU and use that
-            # to compute the image checksum
-            image_hdu = ImageHDU(data=self.data, header=self.header)
-            image_hdu._update_checksum(checksum)
-            if "CHECKSUM" in image_hdu.header:
-                # This will also pass through to the ZHECKSUM keyword and
-                # ZDATASUM keyword
-                self._image_header.set(
-                    "CHECKSUM",
-                    image_hdu.header["CHECKSUM"],
-                    image_hdu.header.comments["CHECKSUM"],
-                )
-            if "DATASUM" in image_hdu.header:
-                self._image_header.set(
-                    "DATASUM",
-                    image_hdu.header["DATASUM"],
-                    image_hdu.header.comments["DATASUM"],
-                )
-            # Store a temporary backup of self.data in a different attribute;
-            # see below
-            self._imagedata = self.data
-
-            # Now we need to perform an ugly hack to set the compressed data as
-            # the .data attribute on the HDU so that the call to _writedata
-            # handles it properly
-            self.__dict__["data"] = self.compressed_data
-
-        return super()._prewriteto(checksum=checksum, inplace=inplace)
-
-    def _writeheader(self, fileobj):
-        """
-        Bypasses `BinTableHDU._writeheader()` which updates the header with
-        metadata about the data that is meaningless here; another reason
-        why this class maybe shouldn't inherit directly from BinTableHDU...
-        """
-        return ExtensionHDU._writeheader(self, fileobj)
+            return self._tmp_bintable._writeto(fileobj, inplace=inplace, copy=copy)
 
-    def _writedata(self, fileobj):
-        """
-        Wrap the basic ``_writedata`` method to restore the ``.data``
-        attribute to the uncompressed image data in the case of an exception.
-        """
-        try:
-            return super()._writedata(fileobj)
-        finally:
-            # Restore the .data attribute to its rightful value (if any)
-            if hasattr(self, "_imagedata"):
-                self.__dict__["data"] = self._imagedata
-                del self._imagedata
-            else:
-                del self.data
+    def _postwriteto(self):
+        del self._tmp_bintable
 
     def _close(self, closed=True):
-        super()._close(closed=closed)
+        # FIXME: need to determine how to keep memmap open if needed
```

Comment:
This comment doesn't make sense, so removed - the reason I had to have this method was because otherwise the default ``_close`` accesses ``self.data`` which would force a decompression. However I think the correct thing to do here is just to pass on the ``_close`` call to the bintable if present, so I am pushing a commit to do that.